### PR TITLE
Features/duo

### PR DESCRIPTION
**Pull request by yzong12138**
bug fixed for not using the segment id for the bert.
bug fixed for concatenating the duobert tokens in a incorrect way
bug NOT fixed yet for DataParallel() writing/reading wrong model in the DistributedHook
add the training procedure for trec_car for the monobert.
add the code for loading the model from huggingface for evaluation use (just a structure, maybe need to be deleted soon). 

start to implement the code for splade.py

### CODE REVIEW

File: xpmir/datasets/adapters.py [hunk 1]
```diff
@@ -71,6 +73,54 @@ def fold(ids: Iterable[str], dataset: Adhoc):
...
+    topics: Annotated[Path, pathgenerator("topics.tsv")]
+    """Generated topics file"""
+
+
+    def config(self) -> Adhoc:
+        return Adhoc(
+            id="",  # No need to have a more specific id since it is generated
+            topics=CSVAdhocTopics(id="", path=self.topics),
+            assessments=TrecAdhocAssessments(id="", path=self.assessments),
+            documents=self.datasets[0].documents,
```
Comment (bpiwowar) [inline xpmir/datasets/adapters.py:96]: please check that all the self.datasets[i].documents are the same (equal id)

File: xpmir/papers/monobert/experiment_car.py [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+        runs_path = xp.resultspath / "runs"
+        cleanupdir(runs_path)
+        runs_path.mkdir(exist_ok=True, parents=True)
+        logging.info("Monitor learning with:")
+        logging.info("tensorboard --logdir=%s", runs_path)
+
+        # Datasets: train, validation and test 
+        documents = prepare_dataset("irds.msmarco-passage.documents") # for indexing 
+        cars_documents = prepare_dataset("irds.car.v1.5.documents") # for indexing
+        # the training dataset used to prepare the pairwise sampler
```
Comment (bpiwowar) [inline xpmir/papers/monobert/experiment_car.py:148]: could be simplified a bit

```py
folds = [prepare_dataset(f"irds.car.v1.5.train.fold{i}") for i in range(4)]

```

File: xpmir/papers/monobert/experiment_car.py [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    req_duration = duration("6 days")
+    launcher = find_launcher(cpu() & req_duration, tags=tags)
+
+    batch_size = batch_size or configuration.Learner.batch_size
+    max_epochs = max_epochs or configuration.Learner.max_epoch
+
+    # FIXME: uses CPU constraints rather than GPU
+    cpu_launcher_4G = find_launcher(cuda_gpu(mem="4G"))
+
+    if configuration.type == "monobert-small" or 'monobert-small-car':
```
Comment (bpiwowar) [inline xpmir/papers/monobert/experiment_car.py:106]: this should be in the `yaml` configuration, not here

File: xpmir/papers/monobert/lr_6e-7.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+type: monobert_lr_6e-7
```
Comment (bpiwowar) [inline xpmir/papers/monobert/lr_6e-7.yaml:1]: no need to have one file per setting / better to allow overriding the default in the experiment file

File: xpmir/papers/monobert/experiment_car.py [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+@click.option(
+    "--port", type=int, default=12345, help="Port for monitoring (default 12345)"
+)
+
+# @omegaconf_argument("configuration", package=__package__)
+# works only with this one a the moment
+@omegaconf_argument("configuration", package="xpmir.papers.monobert")
+@click.argument("workdir", type=Path)
+@click.command()
+def cli(debug, configuration, gpu, tags, host, port, workdir, max_epochs, batch_size):
```
Comment (bpiwowar) [inline xpmir/papers/monobert/experiment_car.py:73]: remove `max_epochs`, `gpu`, `tags`, `batch_size` : this is all in `yaml`

File: xpmir/papers/monobert/evaluation_large.py [hunk 6]
```diff
@@ -0,0 +1,200 @@
+# Implementation of the experiments in the paper
```
Comment (bpiwowar) [inline xpmir/papers/monobert/evaluation_large.py:1]: see other comments

File: xpmir/text/huggingface.py [hunk 7]
```diff
@@ -23,11 +25,39 @@
     logging.error("Install huggingface transformers to use these configurations")
     raise
 
-from xpmir.letor.records import TokenizedTexts
+from xpmir.letor.records import BaseRecords, TokenizedTexts
 import xpmir.text as text
 
 logger = easylog()
 
+class CrossScorerHuggingface(TorchLearnableScorer):
```
Comment (bpiwowar) [inline xpmir/text/huggingface.py:33]: To remove (this will be using huggingface)

**Comment by yzong12138** [comment]
bug fixed for not using the segment id for the bert.
bug fixed for concatenating the duobert tokens in a incorrect way
bug NOT fixed yet for DataParallel() writing/reading wrong model in the DistributedHook
add the training procedure for trec_car for the monobert.
add the code for loading the model from huggingface for evaluation use (just a structure, maybe need to be deleted soon).

start to implement the code for splade.py

**Pull request merged**